### ISSUE TITLE
Infinite loop: mixing watch, useEffect and setState

### ISSUE THREAD

**Issue by rasgo-cc** [react-hook-form/react-hook-form#1515]
**Describe the bug**
Not sure if this is a bug or if I'm missing something. 
I'm using a few Controllers to control custom components (custom inputs) and I want to detect whenever the whole form data changes so I can filter/process it.

Using `watch()` as a dependency of `useEffect` and using state inside of it causes an infinite loop.

**To Reproduce**

https://codesandbox.io/s/react-hook-form-infinite-loop-hx32d?file=/src/App.js

Uncomment the "setFilters()" to reproduce the infinite loop. Without it, it's all good.

**Comment by polovi** [comment]
You use watchAll object in dependency list so the useEffect is called on every render

**Comment by bluebill1049** [comment]
Yes, you should just use watch to return `value` instead of place as dep in the useEffect.

**Comment by heri16** [comment]
@rasgo-cc How was this solved?

**Comment by bluebill1049** [comment]
> @rasgo-cc How was this solved?

what's the issue that you are facing?

**Comment by hosamre94** [comment]
![image](https://user-images.githubusercontent.com/15855930/124792088-57039980-df55-11eb-89aa-0780b8d07e18.png)


I have the same issue here and I can't figure out any good solution to prevent the looping


inside the watch, I should update inputs values depending on some user changes on multiple field's and sometimes I should perform the update also on the changed fields,

I think of two main solutions for this issue, 
the first one is if we can supply a third parameter to setValue for example something like this 
```
setValue ( fieldName , value ,{
     shouldWatch : false
})
```
the other one is if we can watch a specific column inside the fieldArray like if I have something like this `  [{ name, price, total }] ` and only watch for changes on the `total `  field only

or if i can update all the values at once to prevent calling the watch multiple times for each changed input field...!!

**Comment by bluebill1049** [comment]
Watched values doesn't gets memories, that's the issue here. I wouldn't use watch return value as useEffect dep.

**Comment by hosamre94** [comment]
> Watched values doesn't gets memories, that's the issue here. I wouldn't use watch return value as useEffect dep.

for me i don't think this is the case, the useEffect works fine it's only got triggered whenever a value inside the form is changed , but the problem is each `setValue`  inside the useEffect is triggering the `watch` again which call the `setValue` again and so on and so forth i have infinite loop

**Comment by ambewas** [comment]
We're noticing a similar problem - we're using `watch` to access a value and pickup changes to it. The value is from `react-select`  and is thus an object ( `{ value: 'foo', label: 'bar'}` ). 

We're having to set some local state using `useState` depending on which element was selected. So we've got a hook doing something like this: 

```js
const { localState, setLocalState } = useState();
const watchedObject = watch('somePath');

useEffect(() => {
  setLocalState(watchedObject);
}, [watchedObject])
```

in the browser, this runs fine, persumably because `watchedObject` is the same reference everytime -- but for some reason in our specs, this triggers an infinite loop. 

We've solved this with comparing localState with watchedObject -- but it's kinda not ideal to write code just to make specs pass.

```js
const { localState, setLocalState } = useState();
const watchedObject = watch('somePath');

useEffect(() => {
  if (isNotTheSameByDeepCompare(localState, watchedObject)) {
    setLocalState(watchedObject);
  }
}, [watchedObject])
```

Any advice on how to improve this workaround with an official solution would be much appreciated 👍
